head#This script takes the unique list of alleles generated by 'table.R', and pools alleles at the level of sub-genic regions (each locus was divided into five equally sized segments, while requiring a minimum size of 100 bp per segment).

# .. create reference list of genes to monitor progress (clockwise from origin)
linkage<-c('lacZ-1', 'galE-1', 'trpA-1', 'hipA-1', 'cheA-1', 'gyrA-1', 'purL-1', 'argR-1', 'xylA-1', 'uvrA-1')

# .. read the reference file with the genome divided at the level of (3) sub-genic regions
ORFnames<-read.table('parsed_R606genoscope_IUD_n3_ORF.txt', sep="\t", header=FALSE, as.is=TRUE, comment.char = "") #The default is comment.char = "#"!!

# .. read the time series data file
table<-read.table(filename1, sep="\t", header=TRUE, as.is=TRUE, comment.char = "") #The default is comment.char = "#"!!)
table<-table[,c(1:10)]

# .. add column names
colnames(table)<-c('name','alle', 't1',  't2',  't3', 't4', 't5', 'strand', 'srt', 'end')

# .. create vector to store the pooled data
listtot<-c(0,0,0,0,0,0,0,0,0,0,0)

for (orf in ORFnames$V1) {		# loop accross the genome

    list<-c(0,0,0,0,0,0,0,0,0,0,0)	# a recyclable vector to store the pooled data
    alls<-table[table$alle >= ORFnames[ORFnames$V1==orf,2] & table$alle <= ORFnames[ORFnames$V1==orf,3], ]
    if (orf%in%linkage) {print(orf)}	# to monitor progress
    source('pooling.R')		# the pooling script

}

# .. add column names
colnames(listtot)<-c("ORF", "alle", "site", "abn", "t1", "t2", "t3", "t4", "t5","tot", "pos", "strand")

# .. keep the good lines
listtot<-subset(listtot, listtot$alle!=0)

# .. write a time series file with the pooled data
write.table(listtot, file = filename2, sep = "\t")
